File: 1302.层数最深叶子节点的和.cpp
/*
 * @lc app=leetcode.cn id=1302 lang=cpp
 *
 * [1302] 层数最深叶子节点的和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int deepestLeavesSum(TreeNode *root) {
        if (root == nullptr) return 0;

        queue<TreeNode *> qu;
        qu.emplace(root);
        int sum = 0;
        while (!qu.empty()) {
            int sz = qu.size();
            sum = 0;
            while (sz-- > 0) {
                TreeNode *node = qu.front();
                qu.pop();

                sum += node->val;
                if (node->left != nullptr) qu.emplace(node->left);
                if (node->right != nullptr) qu.emplace(node->right);
            }
        }
        return sum;
    }
};
// @lc code=end

File: 1302.层数最深叶子节点的和.java
/*
 * @lc app=leetcode.cn id=1302 lang=java
 *
 * [1302] 层数最深叶子节点的和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    private int maxLevel = 0;
    private int sum = 0;

    public int deepestLeavesSum(TreeNode root) {
        dfs(root, 1);
        return sum;
    }

    public void dfs(TreeNode root, int level) {
        if (root == null) {
            return;
        }

        if (level > maxLevel) {
            maxLevel = level;
            sum = root.val;
        } else if (level == maxLevel) {
            sum += root.val;
        }

        dfs(root.left, level + 1);
        dfs(root.right, level + 1);
    }
}
// @lc code=end
